File: src/api/localApiClient.js
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:4000';

async function apiRequest(method, path, { params = {}, body } = {}) {
  const url = new URL(path, API_BASE_URL);

  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      url.searchParams.set(key, value);
    }
  });

  const init = { method, headers: {} };

  if (body !== undefined) {
    init.headers['Content-Type'] = 'application/json';
    init.body = JSON.stringify(body);
  }

  const res = await fetch(url.toString(), init);
  if (!res.ok) {
    const text = await res.text().catch(() => '');
    let message = text || res.statusText;
    try {
      const json = JSON.parse(text);
      if (json.detail) message = typeof json.detail === 'string' ? json.detail : json.detail.message || message;
    } catch (_) {}
    throw new Error(message);
  }

  // 204 No Content
  if (res.status === 204) {
    return null;
  }
  return res.json();
}

const apiGet = (path, params) => apiRequest('GET', path, { params });
const apiPost = (path, body) => apiRequest('POST', path, { body });
const apiPut = (path, body) => apiRequest('PUT', path, { body });
const apiDelete = (path, params) => apiRequest('DELETE', path, { params });

const actorToPayload = (actor) => ({
  actor_id: actor?.id,
  actor_email: actor?.email,
});

function normalizeWorkoutSession(row) {
  const durationMinutes = row.duration_minutes;
  let parsedDuration = Number(durationMinutes);
  if (Number.isNaN(parsedDuration)) {
    parsedDuration = 0;
  }

  let exercisesCompleted = row.exercises_completed;
  if (typeof exercisesCompleted === 'string') {
    try {
      exercisesCompleted = exercisesCompleted ? JSON.parse(exercisesCompleted) : [];
    } catch {
      exercisesCompleted = [];
    }
  }

  return {
    ...row,
    duration_minutes: parsedDuration,
    exercises_completed: exercisesCompleted,
  };
}

export const localApi = {
  // Auth (SQLite) ------------------------------------------------------------
  async login(email, password) {
    return apiPost('/auth/login', { email, password });
  },

  async register(payload) {
    return apiPost('/auth/register', payload);
  },

  async resetPassword(email, newPassword) {
    return apiPost('/auth/reset-password', { email, new_password: newPassword });
  },

  async getUsers() {
    return apiGet('/auth/users');
  },

  // Workout sessions ---------------------------------------------------------
  async getWorkoutSessionsByUserEmail(userEmail) {
    if (!userEmail) return [];
    const rows = await apiGet('/workout-sessions', { user_email: userEmail });
    return rows.map(normalizeWorkoutSession);
  },

  async createWorkoutSession(data) {
    const row = await apiPost('/workout-sessions', data);
    return normalizeWorkoutSession(row);
  },

  async updateWorkoutSession(id, data) {
    // #region agent log
    fetch('http://127.0.0.1:7878/ingest/edf88b8a-a771-462f-8c17-2acdeb22ffe5', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Debug-Session-Id': 'c7559d',
      },
      body: JSON.stringify({
        sessionId: 'c7559d',
        runId: 'pre-fix',
        hypothesisId: 'H2',
        location: 'localApiClient.js:updateWorkoutSession',
        message: 'Calling apiPut for workout-session',
        data: {
          id,
          keys: Object.keys(data || {}),
        },
        timestamp: Date.now(),
      }),
    }).catch(() => {});
    // #endregion agent log

    const row = await apiPut(`/workout-sessions/${id}`, data);
    return normalizeWorkoutSession(row);
  },

  // Workouts -----------------------------------------------------------------
  async getWorkouts(params) {
    return apiGet('/workouts', params ?? {});
  },

  async getWorkoutById(id) {
    return apiGet(`/workouts/${id}`);
  },

  async createWorkout(data) {
    return apiPost('/workouts', data);
  },

  async updateWorkout(id, data) {
    return apiPut(`/workouts/${id}`, data);
  },

  async deleteWorkout(id) {
    return apiDelete(`/workouts/${id}`);
  },

  // Workout templates ---------------------------------------------------------
  async getWorkoutTemplates(params) {
    return apiGet('/workout-templates', params ?? {});
  },

  async createWorkoutTemplate(data, actor) {
    return apiPost('/workout-templates', { ...data, ...actorToPayload(actor) });
  },

  async updateWorkoutTemplate(id, data, actor) {
    return apiPut(`/workout-templates/${id}`, { ...data, ...actorToPayload(actor) });
  },

  async deleteWorkoutTemplate(id, actor) {
    return apiDelete(`/workout-templates/${id}`, actorToPayload(actor));
  },

  async getTemplateExercisesByTemplateId(templateId) {
    return apiGet(`/workout-templates/${templateId}/exercises`);
  },

  async createTemplateExercise(templateId, data, actor) {
    return apiPost(`/workout-templates/${templateId}/exercises`, { ...data, ...actorToPayload(actor) });
  },

  async updateTemplateExercise(templateId, exerciseId, data, actor) {
    return apiPut(`/workout-templates/${templateId}/exercises/${exerciseId}`, { ...data, ...actorToPayload(actor) });
  },

  async deleteTemplateExercise(templateId, exerciseId, actor) {
    return apiDelete(`/workout-templates/${templateId}/exercises/${exerciseId}`, actorToPayload(actor));
  },

  async assignWorkoutTemplate(studentId, templateId, actor, notes = '', expiresAt = '') {
    return apiPost(`/students/${studentId}/template-assignments`, {
      template_id: templateId,
      notes,
      expires_at: expiresAt || undefined,
      ...actorToPayload(actor),
    });
  },

  async getStudentTemplateAssignments(studentId) {
    return apiGet(`/students/${studentId}/template-assignments`);
  },

  async removeStudentTemplateAssignment(studentId, assignmentId, actor) {
    return apiDelete(`/students/${studentId}/template-assignments/${assignmentId}`, actorToPayload(actor));
  },

  // Exercises ----------------------------------------------------------------
  async getTemplateExercises() {
    return apiGet('/exercises', { is_template: 'true' });
  },

  async getExercisesByWorkoutId(workoutId) {
    if (!workoutId) return [];
    return apiGet('/exercises', { workout_id: workoutId });
  },

  async createExercise(data) {
    return apiPost('/exercises', data);
  },

  async updateExercise(id, data) {
    return apiPut(`/exercises/${id}`, data);
  },

  async deleteExercise(id) {
    return apiDelete(`/exercises/${id}`);
  },

  // Students -----------------------------------------------------------------
  async getStudents(params = {}) {
    const merged = { ...params, with_workout_summary: true };
    return apiGet('/students', merged);
  },

  async createStudent(data) {
    return apiPost('/students', data);
  },

  async getStudent(id) {
    return apiGet(`/students/${id}`);
  },

  async updateStudent(id, data) {
    return apiPut(`/students/${id}`, data);
  },
};

File: flutter_app/lib/api/local_api_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Cliente HTTP simples para falar com o backend Python.
class LocalApiClient {
  LocalApiClient({this.baseUrl = 'http://localhost:4000'});

  final String baseUrl;

  Uri _uri(String path, [Map<String, String>? query]) {
    final uri = Uri.parse('$baseUrl$path');
    if (query == null || query.isEmpty) return uri;
    return uri.replace(queryParameters: {
      ...uri.queryParameters,
      ...query,
    });
  }

  Exception _errorFromResponse(http.Response res) {
    if (res.body.isNotEmpty) {
      return Exception(res.body);
    }
    return Exception(res.reasonPhrase ?? 'Erro na requisição (${res.statusCode})');
  }

  Future<Map<String, dynamic>> _decodeObject(http.Response res) async {
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = res.body.isEmpty ? '{}' : res.body;
      return jsonDecode(body) as Map<String, dynamic>;
    }
    throw _errorFromResponse(res);
  }

  Future<List<dynamic>> _decodeList(http.Response res) async {
    if (res.statusCode >= 200 && res.statusCode < 300) {
      final body = res.body.isEmpty ? '[]' : res.body;
      return jsonDecode(body) as List<dynamic>;
    }
    throw _errorFromResponse(res);
  }

  // -------------------- Auth --------------------

  Future<Map<String, dynamic>> login(String email, String password) async {
    final res = await http.post(
      _uri('/auth/login'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );
    return _decodeObject(res);
  }

  // -------------------- Students --------------------

  Future<List<dynamic>> getStudents() async {
    final res = await http.get(_uri('/students'));
    return _decodeList(res);
  }

  Future<Map<String, dynamic>> createStudent(Map<String, dynamic> data) async {
    final res = await http.post(
      _uri('/students'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _decodeObject(res);
  }

  // -------------------- Workouts --------------------

  Future<List<dynamic>> getWorkoutsForStudent(String studentId) async {
    final res = await http.get(_uri('/workouts', {'student_id': studentId}));
    return _decodeList(res);
  }

  Future<Map<String, dynamic>> createWorkout(Map<String, dynamic> data) async {
    final res = await http.post(
      _uri('/workouts'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _decodeObject(res);
  }

  Future<Map<String, dynamic>> updateWorkout(String id, Map<String, dynamic> data) async {
    final res = await http.put(
      _uri('/workouts/$id'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _decodeObject(res);
  }

  Future<void> deleteWorkout(String id) async {
    final res = await http.delete(_uri('/workouts/$id'));
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw _errorFromResponse(res);
    }
  }

  // -------------------- Workout Sessions --------------------

  Future<Map<String, dynamic>> createQuickWorkoutSession({
    required String userEmail,
    required String workoutId,
    required String workoutName,
  }) async {
    final now = DateTime.now();
    final date = now.toIso8601String().split('T').first;

    final payload = {
      'id': null,
      'user_email': userEmail,
      'workout_id': workoutId,
      'workout_name': workoutName,
      'date': date,
      'start_time': null,
      'end_time': null,
      'duration_minutes': '0',
      'status': 'completed',
      'exercises_completed': '[]',
    };

    final res = await http.post(
      _uri('/workout-sessions'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(payload),
    );
    return _decodeObject(res);
  }

  Future<List<dynamic>> getWorkoutSessionsByUserEmail(String email) async {
    final res = await http.get(_uri('/workout-sessions', {'user_email': email}));
    return _decodeList(res);
  }

  // -------------------- Exercises --------------------

  Future<List<dynamic>> getExercisesForWorkout(String workoutId) async {
    final res = await http.get(_uri('/exercises', {'workout_id': workoutId}));
    return _decodeList(res);
  }

  Future<Map<String, dynamic>> createExercise(Map<String, dynamic> data) async {
    final res = await http.post(
      _uri('/exercises'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _decodeObject(res);
  }

  Future<Map<String, dynamic>> updateExercise(String id, Map<String, dynamic> data) async {
    final res = await http.put(
      _uri('/exercises/$id'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode(data),
    );
    return _decodeObject(res);
  }

  Future<void> deleteExercise(String id) async {
    final res = await http.delete(_uri('/exercises/$id'));
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw _errorFromResponse(res);
    }
  }
}
